radicle-node/src/service.rs

```diff
@@ -1103,18 +1160,23 @@ where
                 }
                 debug!(target: "service", "Fetch accepted for {rid} from {remote}..");
 
-                let namespaces = match self.tracking.namespaces_for(&self.storage, &rid) {
-                    Ok(ns) => ns,
-                    Err(err) => {
-                        if let Some(resp) = self.fetch_reqs.get(&rid) {
-                            resp.send(FetchResult::Failed {
-                                reason: err.to_string(),
-                            })
-                            .ok();
+                let namespaces = if let Some(ns) = self.fetch_namespaces.remove(&timestamp) {
+                    ns
+                } else {
+                    match self.tracking.namespaces_for(&self.storage, &rid) {
+                        Ok(ns) => ns,
+                        Err(err) => {
+                            if let Some(resp) = self.fetch_reqs.get(&rid) {
+                                resp.send(FetchResult::Failed {
+                                    reason: err.to_string(),
+                                })
+                                .ok();
+                            }
+                            return Ok(());
```

Comment:
What is the reason for not always using `tracking.namespaces_for`?

Reply:
The problem with always using `namespaces_for` is that: we are not specifying namespaces based on the namespaces found in the `refs_announcement` message.  We do a [db look up](https://github.com/radicle-dev/heartwood/blob/f8aa1daedb391069847c66e14b2da6cc2e5669ae/radicle-node/src/service/tracking.rs#L118) in local db for trusted nodes.  We are blindly sending all trusted node namespaces found in the db to fetch, as a response of `refs_announcement`. 

For a high level, what I am doing are:

1) we will find the namespaces of [trusted nodes](https://github.com/radicle-dev/heartwood/blob/f055c9d4b38c0472dbc6843dbeeb9f7af8b9b5f0/radicle-node/src/service.rs#L923) from `refs_announcement`. 

2) we will [record the namespaces with its fetch id (timestamp)](https://github.com/radicle-dev/heartwood/blob/f055c9d4b38c0472dbc6843dbeeb9f7af8b9b5f0/radicle-node/src/service.rs#L584-L587) .  This is to simplify the message format, i.e. not to encode a list of namespaces into the message itself.

3) when the `FetchOk` comes back, we will use the id(timestamp) to [look up the namespaces and fetch](https://github.com/radicle-dev/heartwood/blob/f055c9d4b38c0472dbc6843dbeeb9f7af8b9b5f0/radicle-node/src/service.rs#L1163) exactly them.

Reply:
I *think* that makes sense, but even so, I don't see the need to change the `Fetch/FetchOk` messages. There is only one fetch happening at a time per session, so simply storing the namespaces in the session should suffice. As you mentioned, until `FetchOk` is received, we have `Protocol::Gossip { requested: Option<Id> }`, you could simply change that to `requested: Option<(Id, Namespaces)>`.

Reply:
re: `namespaces_for`: what you're saying is that we may want to only fetch a subset of namespaces, since the announcing node may not have some of our trusted namespaces?

I think it might be ok to let Git handle that, ie. the refs that aren't found on the other side will just be ignored.. This would simplify the service code at almost no cost? cc @FintanH

Reply:
> re: `namespaces_for`: what you're saying is that we may want to only fetch a subset of namespaces, since the announcing node may not have some of our trusted namespaces?
> 
> I think it might be ok to let Git handle that, ie. the refs that aren't found on the other side will just be ignored.. This would simplify the service code at almost no cost? cc @FintanH

Ya, this is correct :point_up: We just say, "Hey give me the things I'm interested in", if these don't exist then we don't get them and that's fine.

Reply:
> This would simplify the service code at almost no cost?

The trusted node list from local db could be much larger than the trusted nodes in `refs_announcements`. For example, there might be only one trusted ref in the announcement. In practice, it would depend on how radicle-node scales.

Again, like I said earlier, it is doable to blindly fetch using trusted nodes list from the local db. I think it's less ideal, and also defeats the purpose of `refs_announcement` because now the `refs_announcement` will behave like this: 

1. Announcer sends `refs_announcement` with a list of refs.
2. Receiver sends back `Fetch<rid>` .
3. Announcer sends  back `FetchOk<rid>`.
4. Receiver sends git fetch with all trusted namespaces from local db.
5. Announcer sends back namespaces it has for trusted nodes. 

It seems to me that the 1st step could just be `inventory_announcement` as well. 

For now, I will check the list of refs in `refs_announcement` anyway, make sure there is at least one trusted ref.

Reply:
Yeah, I think you have a point in that it's maybe not ideal for large trusted set. Note that as you said, we should anyway check that at least one ref is interesting before requesting a fetch, and also we're able to check that the `oid` of the ref is one we do not already have (currently we don't look at ref oids), which is not something we can do with inventory announcements.

I'd be supportive in storing the matching namespace (basically an intersection between trusted and announced) as part of the session state, to be able to only request what the other peer has.

Reply:
> I'd be supportive in storing the matching namespace (basically an intersection between trusted and announced) as part of the session state, to be able to only request what the other peer has.

I've looked into that. Seems to me it has a problem: when we queue pending fetches, we will not be able to obtain this namespace info for a protocol state update when it's time to send out the next pending fetch. Unless we store the info with each fetch, which is back to the original approach.

Reply:
I've refactored this patch :

1. Removed `Timestamp` in the `Fetch` message.

2. Re-use `namespaces_for()` to get the namespaces for a fetch. Then use them to check against the refs in `refs_announcement`.

3. Fixed `namespace_for()` to support `scope::Trusted` (merged from https://github.com/radicle-dev/heartwood/pull/420) . I understand @FintanH provided comments over there, but I wanted to ask for another review in this context. My point is that we should fetch from the initial remote / seed when the repo didn't exist locally yet.